tests/test_http_request.py

```diff
@@ -1147,5 +1148,57 @@ def test_latin1(self):
         self._test_request(params=(u'pas£',), encoding='latin1')
 
 
+class JSONRequestTest(RequestTest):
+    request_class = JSONRequest
+    default_method = 'POST'
+    default_headers = {b'Content-Type': [b'application/json']}
+
+    def test_body(self):
+        r1 = self.request_class(url="http://www.example.com/")
+        self.assertEqual(r1.body, b'{}')
+
+        r2 = self.request_class(url="http://www.example.com/", body=b"")
+        self.assertEqual(r2.body, b'{}')
+
+        data = {
+            'name': 'value',
+        }
+        r3 = self.request_class(url="http://www.example.com/", data=data)
+        self.assertEqual(r3.body, to_bytes(json.dumps(data)))
+
+        r4 = self.request_class(url="http://www.example.com/", body='body1', data=data)
+        self.assertEqual(r3.body, to_bytes(json.dumps(data)))
+
+    def test_replace(self):
+        """Test Request.replace() method"""
+        r1 = self.request_class("http://www.example.com")
+        hdrs = Headers(r1.headers)
+        hdrs[b'key'] = b'value'
+        r2 = r1.replace(body="New body", headers=hdrs)
```

Comment:
For ``XmlRpcRequest``, this would replace the body, and for me it would look surprising that the body value would be ignored. What is the advantage of such behaviour? Here is how ``XmlRpcRequest`` handles this in ``__init__``: https://github.com/scrapy/scrapy/blob/886513c3751b92e42dcc8cb180d4c15a5a11ccaf/scrapy/http/request/rpc.py#L20
With such implementation it seems a custom ``replace`` method is not needed, and behaviour is less surprising.